### PR TITLE
ENT-1300 | More robust date sorting in tables

### PR DESCRIPTION
**Pull request by christopappas**
...now with tests!

### CODE REVIEW

File: src/data/actions/table.js [hunk 1]
```diff
@@ -108,11 +88,54 @@ const sortTable = (tableId, fetchMethod, ordering) => (
     };
     dispatch(sortRequest(tableId, ordering));
 
+    const isDesc = ordering.startsWith('-');
+    const orderField = isDesc ? ordering.substring(1) : ordering;
+    // customSort relies on orderField. Should orderField be handed to customSort or is this enough?
```
Comment (christopappas) [inline src/data/actions/table.js]: check my comment please :)

File: src/components/EnrollmentsTable/EnrollmentsTable.mocks.js [hunk 2]
```diff
@@ -5,109 +5,132 @@ const mockEnrollmentFetchResponse = {
...
     }, {
-      id: 150, user_email: 'Bettie.Kub90@bestrun.com', course_title: 'Foundations of Data Science: Computational Thinking with Python', course_price: 99, course_start: '2018-04-12T03:10:04.321Z', course_end: '2018-10-25T08:11:55.481Z', passed_timestamp: '2018-10-04T17:04:52.362Z', has_passed: true, current_grade: 0.76, last_activity_date: '2018-09-04T07:40:06.098Z',
+      id: 150, user_email: 'Bettie.Kub90@bestrun.com', course_title: 'Foundations of Data Science: Computational Thinking with Python', course_price: '99.00', course_start: '2018-04-12T03:10:04.321Z', course_end: '2018-10-25T08:11:55.481Z', passed_timestamp: '2018-10-04T17:04:52.362Z', has_passed: true, current_grade: 0.76, last_activity_date: '2018-09-04T07:40:06.098Z',
     }, {
       id: 239, user_email: 'Bill_Friesen26@bestrun.com', course_title: 'Reputation Management in a Digital World', course_price: 99, course_start: '2017-12-06T18:14:22.647Z', course_end: '2018-05-21T23:28:21.493Z', passed_timestamp: '2018-02-27T00:27:31.010Z', has_passed: true, current_grade: 0.86, last_activity_date: '2018-08-28T08:48:27.389Z',
```
[the remark this answers is not in the document]
Comment (christopappas) [reply src/components/EnrollmentsTable/EnrollmentsTable.mocks.js]: Fixed in latest commit!

File: src/data/actions/table.js [hunk 3]
```diff
@@ -80,23 +79,38 @@ const paginateTable = (tableId, fetchMethod, pageNumber) => (
...
+    let b = value2[orderField] || '';
+    // if both are strings
+    if (typeof a === 'string' && typeof b === 'string') {
+      // if both are dates
+      if (!Number.isNaN(Date.parse(a)) && !Number.isNaN(Date.parse(b))) {
+        a = new Date(a).getTime();
+        b = new Date(b).getTime();
+        return a - b;
+      }
+      // if A is a date and B is not
```
[the remark this answers is not in the document]
Comment (christopappas) [reply src/data/actions/table.js]: Updated in latest commit!

File: src/data/actions/table.js [hunk 4]
```diff
@@ -80,23 +79,38 @@ const paginateTable = (tableId, fetchMethod, pageNumber) => (
...
+      if (!Number.isNaN(Date.parse(a)) && Number.isNaN(Date.parse(b))) {
+        return 1;
+      }
+      // if B is a date and A is not
```
[the remark this answers is not in the document]
Comment (christopappas) [reply src/data/actions/table.js]: Updated in latest commit!

File: src/data/actions/table.js [hunk 5]
```diff
@@ -110,9 +124,13 @@ const sortTable = (tableId, fetchMethod, ordering) => (
 
     // If we can sort client-side because we have all of the data, do that
     if (tableState.data && tableState.data.num_pages === 1) {
+      const isDesc = ordering.startsWith('-');
+      const orderField = isDesc ? ordering.substring(1) : ordering;
+      let result = customSort(tableState.data.results, orderField);
```
[the remark this answers is not in the document]
Comment (christopappas) [reply src/data/actions/table.js]: Updated in latest commit!

File: src/data/actions/table.test.js [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+          },
+          fetchMethod,
+          ordering,
+        },
+      });
+
+      store.dispatch(sortTable(tableId, fetchMethod, ordering));
+
+      let actual = store.getActions()[1].payload.data.results;
+      expect(null).toEqual(actual[0].last_activity_date);
```
[the remark this answers is not in the document]
Comment (christopappas) [reply src/data/actions/table.test.js]: Update in latest commit!

File: src/data/actions/table.test.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+      ordering = '-last_activity_date';
+      store.dispatch(sortTable(tableId, fetchMethod, ordering));
+
+      actual = store.getActions()[3].payload.data.results;
+      expect(null).toEqual(actual.slice(-1)[0].last_activity_date);
+      expect('2018-09-13T17:36:05.725Z').toEqual(actual[0].last_activity_date);
+    });
+
+    it('sorts tables based on the current grade', () => {
```
[the remark this answers is not in the document]
Comment (christopappas) [reply src/data/actions/table.test.js]: Noted 👍 -- will need to hand this off to you, as I'm running short on time

File: src/data/actions/table.js [hunk 8]
```diff
@@ -80,23 +79,35 @@ const paginateTable = (tableId, fetchMethod, pageNumber) => (
...
-  // `parseKeyValue` adjusts the key's value into its appropriate data type to ensure
-  // proper sorting and sort order (e.g., asc/desc). A numeric value (even if passed in
-  // as a string) must be parsed as an actual numeric value. An empty value (e.g., null,
-  // undefined) must be parsed as an empty string to ensure the empty values are forced
-  // to the top in an ascending sort order.
-  const parseKeyValue = (obj) => {
-    const value = obj[column] || '';
-    if (!Number.isNaN(value) && !Number.isNaN(parseFloat(value))) {
-      return parseFloat(value);
+const sortBy = (dataToSort, orderField) => {
```
Comment (georgebabey) [inline src/data/actions/table.js]: This is probably fine for now but we may want to move it to utils as I can see it being used in other action handlers in the future.

Comment (adamstankiewicz) [reply src/data/actions/table.js]: Agreed 👍

**Pull request merged**